pkg/querier/queryrange/split_by_interval.go

```diff
@@ -8,77 +8,143 @@ import (
+func SplitByIntervalMiddleware(interval time.Duration, limits queryrange.Limits, merger queryrange.Merger) queryrange.Middleware {
 	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
-		return splitByInterval{
-			next:      next,
-			limits:    limits,
-			merger:    merger,
-			interval:  interval,
-			batchSize: batchSize,
+		return &splitByInterval{
+			next:     next,
+			limits:   limits,
+			merger:   merger,
+			interval: interval,
 		}
 	})
 }
 
+type lokiResult struct {
+	req  queryrange.Request
+	resp chan queryrange.Response
+	err  chan error
+}
+
 type splitByInterval struct {
-	next      queryrange.Handler
-	limits    queryrange.Limits
-	merger    queryrange.Merger
-	interval  time.Duration
-	batchSize int
+	next     queryrange.Handler
+	limits   queryrange.Limits
+	merger   queryrange.Merger
+	interval time.Duration
 }
 
-func (s splitByInterval) Do(ctx context.Context, r queryrange.Request) (queryrange.Response, error) {
-	lokiRequest := r.(*LokiRequest)
-	intervals := splitByTime(lokiRequest, s.interval)
-	var result queryrange.Response
+func (h *splitByInterval) Feed(ctx context.Context, input []*lokiResult) chan *lokiResult {
+	ch := make(chan *lokiResult)
 
-	if lokiRequest.Direction == logproto.BACKWARD {
-		for i, j := 0, len(intervals)-1; i < j; i, j = i+1, j-1 {
-			intervals[i], intervals[j] = intervals[j], intervals[i]
+	go func() {
+		defer close(ch)
+		for _, d := range input {
+			select {
+			case <-ctx.Done():
+				return
+			case ch <- d:
+				continue
+			}
 		}
+	}()
+
+	return ch
+}
+
+func (h *splitByInterval) Process(
+	ctx context.Context,
+	parallelism int,
+	threshold int64,
+	input []*lokiResult,
+) (responses []queryrange.Response, err error) {
+	ch := h.Feed(ctx, input)
+
+	// queries with 0 limits should not be exited early
+	var unlimited bool
+	if threshold == 0 {
+		unlimited = true
 	}
 
-	for _, interval := range intervals {
-		sp, ctx := opentracing.StartSpanFromContext(ctx, "splitByInterval.interval")
-		linterval := interval.(*LokiRequest)
-		logRequest(sp, linterval)
-		reqs := splitByTime(linterval, linterval.EndTs.Sub(linterval.StartTs)/time.Duration(s.batchSize))
+	for i := 0; i < parallelism; i++ {
+		go h.loop(ctx, ch)
+	}
 
-		reqResps, err := queryrange.DoRequests(ctx, s.next, reqs, s.limits)
-		if err != nil {
-			sp.Finish()
+	for _, x := range input {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-x.err:
 			return nil, err
-		}
+		case resp := <-x.resp:
 
-		resps := make([]queryrange.Response, 0, len(reqResps))
-		if result != nil {
-			resps = append(resps, result)
-		}
-		for _, reqResp := range reqResps {
-			resps = append(resps, reqResp.Response)
+			responses = append(responses, resp)
+
+			// see if we can exit early if a limit has been reached
+			if !unlimited {
+				threshold -= resp.(*LokiResponse).Count()
```

Comment:
Not sure about this, are the ordering of responses preserved when checking this ?

Reply:
yup